### ISSUE TITLE
tridentctl image requires a cluster to run

### ISSUE THREAD

**Issue by tonyskapunk** [NetApp/trident#675]
**Describe the bug**
When running `tridentctl images` I expect it to print the images required but fails if a cluster is not available

**Environment**
Provide accurate information about the environment to help us reproduce the issue.

- Other:
  - Tridentctl version: v21.10.0

**To Reproduce**
Run `tridentclt images` without a `KUBECONFIG`

```
KUBECONFIG=/dev/null ./tridentctl --debug images
FATA[0000] could not initialize Kubernetes client; invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
```

**Expected behavior**
Being able to obtain the list of images used by trident without requiring a cluster.

**Additional context**

**Comment by gnarl** [comment]
Hi @tonyskapunk,

The tridentctl command does need to inspect the cluster to determine the correct CSI snapshotter sidecar version to report. We expect that there will be additional checks in the future that also require access to the K8S cluster to accurately report the versions of container images that are required.

If you want a quick reference of the images that are required for a particular Kubernetes version you can check the Trident [requirements documentation](https://docs.netapp.com/us-en/trident/trident-get-started/requirements.html#container-images-and-corresponding-kubernetes-versions).

**Comment by tonyskapunk** [comment]
Hello @gnarl thanks for the response and the link to the docs, that's quite helpful.

Do you know if that inspection is due to the k8s cluster version in question? or if there's any additional information from the cluster that is needed? I'm asking because I see `tricentctl` allows what seems the definition of the k8s version: `--k8s-version`

The use case in mind is a disconnected cluster (no internet access), where it would be nice to get the list of images needed by trident before installing the cluster so I can mirror those required images beforehand.

**Comment by wonderland** [comment]
*ping* 

It would be really nice if this works without already having access to a k8s cluster. The typical use case for `tridentctl images` is to get the required images while still being outside the secure/air-gapped K8s environment. 

The docs are not updated for patch releases (e.g. even if you download 21.10.1, the docs still list the 21.10.0 images). Therefore getting the correct and up-to-date list via tridentctl is the better option, imho. The k8s version can be easily provided via the existing flag. Is there anything else tridentctl would require that drives the need for cluster access?

**Comment by gnarl** [comment]
@wonderland and @tonyskapunk, access to the K8S cluster is needed to determine the correct version of the CSI snapshotter sidecar that is needed. The snapshotter version is determined based on the volumesnapshots.snapshot.storage.k8s.io CRD's version (v1, v1beta1).

**Comment by khassel** [comment]
All these image info's can be found in `cli/k8s_client/yaml_factory.go`.

Executing `cat yaml_factory.go | sed -n 's|.*{CSI_SIDECAR_REGISTRY}/||gIp'  | sort | uniq` will bring this result:
```bash
csi-attacher:v3.4.0
csi-node-driver-registrar:v2.4.0
csi-provisioner:v2.2.2
csi-provisioner:v3.1.0
csi-resizer:v1.3.0
csi-snapshotter:{CSI_SNAPSHOTTER_VERSION}
```

and as mentioned by @gnarl the csi-snapshotter tag depends on v1/v1beta1.

In current `master` the tag could be "v3.0.3" or "v5.0.0" (which is hardcoded in yaml_factory.go).

So why not deliver both images as result of `tridentctl images`? Or make a new proc `tridentctl images-all`

I think most users need this list to bring the images from outside into an internal docker registry so no one matters if there is one image in this list which is not used.

**Comment by khassel** [comment]
another approach could be to add a new flag e.g. `--snap-crd-version`  to `tridentctl images` and if this is provided the cluster access is omitted.

If you run the current implementation against a cluster where no snapshotter is installed or the serviceaccount has no rights to read crd's `tridentctl images` always falls back to the old crd version.

One use case to automate complete installation in a new cluster would be
- run `tridentctl images -v $(kubectl version -o json | jq -r '.serverVersion.gitVersion') -o json | jq -r '.imageSets[0].images[]'` to get images needed
- install external csi-snapshotter with version extracted from the image (e.g. `k8s.gcr.io/sig-storage/csi-snapshotter:v5.0.0`)
- install trident

With current implementation you would install the wrong (old) snapshotter.

**Comment by scaleoutsean** [comment]
`tridentctl version` also requires a cluster to run (in addition to not actually giving version info about the binary `tridentctl` itself even when it runs online). 

That doesn't mean it should run offline. But something like `--offline` could be used for both image and version info.

**Comment by adkerr** [comment]
@Scaleoutsean, if you use the `--client` command it should work without a cluster and `-o yaml` should give additional information like go version for the binary

**Comment by scaleoutsean** [comment]
Thanks, @adkerr! In that case `--client` would work for me if it was a way to get a list of supported `images`.

**Comment by tonyskapunk** [comment]
Hello, I wonder if the `tridentctl images`  still requires to dynamically calculate the csi snapshotter image on the newer versions, looked at 24.06

https://github.com/NetApp/trident/blob/cb68cb389d9dea65eff2d2509b4784827bfbfe5c/cli/k8s_client/yaml_factory.go#L699

And seems like the version is fixed

Oh! And looks like v1betav1 has been [deprecated in v23.07](https://github.com/NetApp/trident/releases/tag/v23.07.0) hopefully this would remove the need to connect to a k8s cluster to print the required images :smile_cat:

**Comment by dmc5179** [comment]
In my experience a running cluster hasn't been needed. There's a doc that lists them:
https://docs.netapp.com/us-en/trident/trident-get-started/requirements.html#storage-system-configuration

The biggest challenge I've faced with disconnected operations is that the helm chart itself doesn't reference these other images. This means that one needs to look to the docs page to ensure all required images are transferred to an offline environment. 

Helm provides for a "spec.relatedImages" section. If these related images were added to the helm chart, it would make mirroring for offline usage much easier.

**Comment by sjpeeris** [comment]
@tonyskapunk We may get to this in a future release. We have created an engineering tracking issue on our side. Closing this.

**Comment by khassel** [comment]
I see no `--client` option for `tridentctl images`, only for `tridentctl version`

```bash
~❯ tridentctl version --client
+----------------+
| CLIENT VERSION |
+----------------+
| 24.10.0        |
+----------------+

~❯ tridentctl images --client
Error: unknown flag: --client

~❯ tridentctl images
FATA[0000] could not initialize Kubernetes client; found the Kubernetes CLI, but it exited with error: The connection to the server localhost:8080 was refused - did you specify the right host or port?

~❯ tridentctl images --help
Print a table of the container images Trident needs

Usage:
  tridentctl images [flags]

Flags:
  -h, --help                 help for images
  -v, --k8s-version string   Semantic version of Kubernetes cluster (default "all")

Global Flags:
  -d, --debug               Set the log level to debug
  -k, --kubeconfig string   Kubernetes config path
      --log-level string    Log level (trace, debug, warn, info, error, fatal (default "info")
  -n, --namespace string    Namespace of Trident deployment
  -o, --output string       Output format. One of json|yaml|name|wide|ps (default)
  -s, --server string       Address/port of Trident REST interface (127.0.0.1 or [::1] only)
```

**Comment by sjpeeris** [comment]
@khassel  you are right. We will create an engineering ticket on our side to address this in a future release. Thanks !!

**Comment by tonyskapunk** [comment]
Thanks for looking into it @sjpeeris Should this one be reopened so people is aware on when that functionality will be available?

**Comment by Khalid-NetApp** [comment]
checking in for an update on this — running into the same pain point in air-gapped/private-registry deployments. Since tridentctl images requires kubectl and a reachable cluster context just to resolve the CSI snapshotter sidecar version, it can't be used purely "pre-install" (before the target K8s environment or a mirrored registry exists). Combined with docs.netapp.com only publishing image lists for major releases (not minor/patch), there's currently no clean way to get an accurate image list for a minor release without either a live cluster. Any plans to decouple the snapshotter-version check from live cluster access, or to have tridentctl images fall back to a documented default when no cluster is reachable? Would help a lot for planning image mirroring ahead of installs. Thanks!
